File: Version4/backend/app/services/prediction_service.py
"""
Prediction service for DCGAN weather predictions
"""

import asyncio
import numpy as np
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Any
import random

from app.core.database import get_db


class PredictionService:
    """Service for managing DCGAN weather predictions"""
    
    def __init__(self):
        self.db = get_db()
        self.model_loaded = False
    
    async def get_predictions(self, time_horizon: int, region: str, prediction_type: str) -> Dict[str, Any]:
        """Get weather predictions with filters"""
        
        # Mock predictions data
        predictions = []
        for i in range(min(time_horizon, 24)):  # Limit to 24 predictions max
            timestamp = datetime.now() + timedelta(hours=i+1)
            confidence = 0.85 + random.random() * 0.14  # 85-99% confidence
            
            predictions.append({
                "id": i + 1,
                "imageUrl": f"https://picsum.photos/seed/pred_{region}_{i}/200/150.jpg",
                "highResUrl": f"https://picsum.photos/seed/pred_hr_{region}_{i}/800/600.jpg",
                "timeLabel": f"Hour {i + 1}" if prediction_type == "hourly" else f"Day {i + 1}",
                "timestamp": timestamp.isoformat(),
                "confidence": round(confidence, 3),
                "region": region,
                "timeHorizon": time_horizon
            })
        
        # Mock loss data for charts
        loss_data = {
            "labels": [f"Epoch {i}" for i in range(1, 6)],
            "datasets": [
                {
                    "label": "Generator Loss",
                    "data": [0.5 - i*0.05 for i in range(5)],
                    "borderColor": "rgb(75, 192, 192)",
                    "backgroundColor": "rgba(75, 192, 192, 0.2)",
                    "tension": 0.1
                },
                {
                    "label": "Discriminator Loss",
                    "data": [0.6 - i*0.08 for i in range(5)],
                    "borderColor": "rgb(255, 99, 132)",
                    "backgroundColor": "rgba(255, 99, 132, 0.2)",
                    "tension": 0.1
                }
            ]
        }
        
        return {
            "predictions": predictions,
            "lossData": loss_data,
            "total": len(predictions),
            "region": region,
            "timeHorizon": time_horizon,
            "type": prediction_type
        }
    
    async def get_recent_predictions(self, limit: int = 10) -> List[Dict[str, Any]]:
        """Get recent predictions"""
        
        predictions = []
        for i in range(min(limit, 10)):
            timestamp = datetime.now() - timedelta(minutes=i*15)
            predictions.append({
                "id": i + 1,
                "location": random.choice(["North America", "Europe", "Asia", "Africa", "South America"]),
                "timestamp": timestamp.strftime("%Y-%m-%d %H:%M:%S"),
                "accuracy": random.randint(85, 98),
                "status": random.choice(["completed", "processing", "completed"])
            })
        
        return predictions
    
    async def generate_prediction_async(self, time_horizon: int, region: str, prediction_type: str):
        """Generate new prediction asynchronously"""
        
        # Simulate DCGAN generation time
        await asyncio.sleep(2)
        
        # Generate mock prediction
        prediction_id = random.randint(1000, 9999)
        confidence = 0.85 + random.random() * 0.14
        
        # Store in database
        await self.db.execute_query(
            """INSERT INTO predictions 
               (image_url, confidence, time_horizon, region, prediction_type, status) 
               VALUES (?, ?, ?, ?, ?, ?)""",
            (
                f"https://picsum.photos/seed/new_pred_{prediction_id}/200/150.jpg",
                confidence,
                time_horizon,
                region,
                prediction_type,
                "completed"
            )
        )
        
        print(f"✅ Generated prediction {prediction_id} for {region}")
    
    async def get_prediction_by_id(self, prediction_id: int) -> Optional[Dict[str, Any]]:
        """Get specific prediction by ID"""
        
        # Mock prediction data
        return {
            "id": prediction_id,
            "imageUrl": f"https://picsum.photos/seed/pred_{prediction_id}/400/300.jpg",
            "highResUrl": f"https://picsum.photos/seed/pred_hr_{prediction_id}/1200/900.jpg",
            "timestamp": datetime.now().isoformat(),
            "confidence": 0.92,
            "region": "global",
            "timeHorizon": 24,
            "type": "hourly",
            "status": "completed"
        }
    
    async def get_confidence_scores(self, prediction_id: int) -> Dict[str, float]:
        """Get confidence scores for a prediction"""
        
        return {
            "overall": round(0.85 + random.random() * 0.14, 3),
            "cloudDetection": round(0.88 + random.random() * 0.11, 3),
            "weatherPattern": round(0.82 + random.random() * 0.17, 3),
            "temporalConsistency": round(0.90 + random.random() * 0.09, 3)
        }
    
    async def delete_prediction(self, prediction_id: int) -> bool:
        """Delete a prediction"""
        
        # Mock deletion
        await self.db.execute_query("DELETE FROM predictions WHERE id = ?", (prediction_id,))
        return True
    
    async def get_stats(self) -> Dict[str, Any]:
        """Get prediction statistics"""
        
        return {
            "accuracy": 94.5,
            "totalPredictions": 1247,
            "avgGenTime": 2.3,
            "modelVersion": "4.0",
            "lastTraining": "2 hours ago",
            "trainingData": "60,000+ frames",
            "successRate": 98.2
        }
    
    async def health_check(self) -> str:
        """Health check for prediction service"""
        
        return "healthy"
    
    async def generate_sample_predictions(self):
        """Generate sample predictions for testing"""
        
        regions = ["global", "north-america", "europe", "asia"]
        types = ["hourly", "daily", "weekly"]
        
        for region in regions:
            for pred_type in types:
                for i in range(5):
                    await self.db.execute_query(
                        """INSERT INTO predictions 
                           (image_url, confidence, time_horizon, region, prediction_type, status) 
                           VALUES (?, ?, ?, ?, ?, ?)""",
                        (
                            f"https://picsum.photos/seed/sample_{region}_{pred_type}_{i}/200/150.jpg",
                            round(0.85 + random.random() * 0.14, 3),
                            random.choice([1, 6, 24, 72, 168]),
                            region,
                            pred_type,
                            "completed"
                        )
                    )
        
        print("✅ Sample prediction data generated")

File: Version4/frontend/src/services/predictionService.js
import axios from 'axios';

const API_BASE_URL = process.env.REACT_APP_API_URL || 'http://localhost:8000';

const predictionService = {
  // Get predictions
  getPredictions: async (params) => {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/predictions`, { params });
      return response.data;
    } catch (error) {
      console.error('Error fetching predictions:', error);
      // Return mock data
      return {
        predictions: [
          {
            id: 1,
            imageUrl: 'https://picsum.photos/seed/pred1/200/150.jpg',
            highResUrl: 'https://picsum.photos/seed/pred1hr/800/600.jpg',
            timeLabel: 'Hour 1',
            timestamp: '2024-01-20 15:00:00',
            confidence: 0.94
          },
          {
            id: 2,
            imageUrl: 'https://picsum.photos/seed/pred2/200/150.jpg',
            highResUrl: 'https://picsum.photos/seed/pred2hr/800/600.jpg',
            timeLabel: 'Hour 2',
            timestamp: '2024-01-20 16:00:00',
            confidence: 0.91
          },
          {
            id: 3,
            imageUrl: 'https://picsum.photos/seed/pred3/200/150.jpg',
            highResUrl: 'https://picsum.photos/seed/pred3hr/800/600.jpg',
            timeLabel: 'Hour 3',
            timestamp: '2024-01-20 17:00:00',
            confidence: 0.88
          }
        ],
        lossData: {
          labels: ['Epoch 1', 'Epoch 2', 'Epoch 3', 'Epoch 4', 'Epoch 5'],
          datasets: [
            {
              label: 'Generator Loss',
              data: [0.5, 0.4, 0.35, 0.3, 0.25],
              borderColor: 'rgb(75, 192, 192)',
              backgroundColor: 'rgba(75, 192, 192, 0.2)',
              tension: 0.1
            },
            {
              label: 'Discriminator Loss',
              data: [0.6, 0.55, 0.45, 0.35, 0.2],
              borderColor: 'rgb(255, 99, 132)',
              backgroundColor: 'rgba(255, 99, 132, 0.2)',
              tension: 0.1
            }
          ]
        }
      };
    }
  },

  // Generate new prediction
  generatePrediction: async (params) => {
    try {
      const response = await axios.post(`${API_BASE_URL}/api/predictions/generate`, params);
      return response.data;
    } catch (error) {
      console.error('Error generating prediction:', error);
      // Return mock prediction
      return {
        prediction: {
          id: Date.now(),
          imageUrl: `https://picsum.photos/seed/newpred${Date.now()}/200/150.jpg`,
          highResUrl: `https://picsum.photos/seed/newpredhr${Date.now()}/800/600.jpg`,
          timeLabel: `Hour ${params.timeHorizon || 1}`,
          timestamp: new Date().toLocaleString(),
          confidence: 0.89 + Math.random() * 0.1
        }
      };
    }
  },

  // Get model statistics
  getModelStats: async () => {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/model/stats`);
      return response.data;
    } catch (error) {
      console.error('Error fetching model stats:', error);
      // Return mock data
      return {
        accuracy: 94.5,
        totalPredictions: 1247,
        avgGenTime: 2.3,
        modelVersion: '4.0',
        lastTraining: '2 hours ago',
        trainingData: '60,000+ frames'
      };
    }
  },

  // Get training history
  getTrainingHistory: async () => {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/model/training-history`);
      return response.data;
    } catch (error) {
      console.error('Error fetching training history:', error);
      return [];
    }
  },

  // Get model performance metrics
  getModelPerformance: async (timeRange) => {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/model/performance`, {
        params: { timeRange }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching model performance:', error);
      return {
        accuracy: 94.5,
        loss: 0.05,
        precision: 0.92,
        recall: 0.89,
        f1Score: 0.90
      };
    }
  },

  // Retrain model
  retrainModel: async (params) => {
    try {
      const response = await axios.post(`${API_BASE_URL}/api/model/retrain`, params);
      return response.data;
    } catch (error) {
      console.error('Error retraining model:', error);
      throw error;
    }
  },

  // Get prediction confidence scores
  getConfidenceScores: async (predictionId) => {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/predictions/${predictionId}/confidence`);
      return response.data;
    } catch (error) {
      console.error('Error fetching confidence scores:', error);
      return {
        overall: 0.89,
        cloudDetection: 0.92,
        weatherPattern: 0.87,
        temporalConsistency: 0.91
      };
    }
  }
};

export default predictionService;
